File: java/src/main/java/de/sven_torben/serialization_benchmark/logging/CsvFileLogger.java
package de.sven_torben.serialization_benchmark.logging;

import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;

import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.StringUtils;

public final class CsvFileLogger extends AbstractDetailsLogger {

	private File logFile;

	public CsvFileLogger() {
		this(Paths.get(System.getProperty("user.dir"), "logs", "benchmark.log")
				.toAbsolutePath().toString());
	}

	public CsvFileLogger(final String logFilePath) {
		if (StringUtils.isNotEmpty(logFilePath)) {
			this.logFile = new File(logFilePath);
			ensureCleanLogFile();
			System.out.println(String.format("CSV Log: %s",
					logFile.getAbsolutePath()));
		}
	}

	private void ensureCleanLogFile() {
		if (this.logFile.exists()) {
			this.logFile.delete();
		}
		try {
			FileUtils.touch(logFile);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	@Override
	protected final void log(final String serializerName, final int bytes,
			final int iterations, final long minWrite, final long minRead,
			final double avgWrite, final double avgRead, final long maxWrite,
			final long maxRead) {
		if (logFile != null) {
			try {
				FileUtils.write(logFile, String.format(
						"%s;%d;%d;%d;%d;%.2f;%.2f;%d;%d%n", serializerName,
						bytes, iterations, minWrite, minRead, avgWrite,
						avgRead, maxWrite, maxRead), true);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

}

File: net/Logging/CsvFileLogger.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace De.Sven_Torben.Serialization_Benchmark.Logging
{
    sealed class CsvFileLogger : AbstractDetailsLogger
    {
        private static readonly Encoding ENCODING = Encoding.UTF8;

        private string logFile;

        public CsvFileLogger() : this(Path.Combine(Directory.GetCurrentDirectory(), "logs", "benchmark.log")) { }

        public CsvFileLogger(string logFilePath)
        {
            if (!String.IsNullOrEmpty(logFilePath))
            {
                this.logFile = logFilePath;
                EnsureCleanLogfile();
                Console.Out.WriteLine("CSV Log: {0}", logFile);
            }
        }

        private void EnsureCleanLogfile()
        {
            if (File.Exists(logFile))
            {
                File.Delete(logFile);
            }
            string dirName = Path.GetDirectoryName(logFile);
            if (!Directory.Exists(dirName))
            {
                Directory.CreateDirectory(dirName);
            }
            File.AppendAllText(logFile, "serializerName;bytes;iterations;min write;min read;avg write;avg read; max write; max read" + Environment.NewLine, ENCODING);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        protected override void Log(string serializerName, int bytes, int iterations, long minWrite, long minRead, double avgWrite, double avgRead, long maxWrite, long maxRead)
        {
            if (!String.IsNullOrEmpty(logFile))
            {
                File.AppendAllText(logFile, string.Format("{0};{1};{2};{3};{4};{5};{6};{7};{8}{9}", serializerName, bytes, iterations, minWrite, minRead, avgWrite, avgRead, maxWrite, maxRead, Environment.NewLine), ENCODING);
            }
        }
    }
}
